### PR TITLE
588 read reply

### PR DESCRIPTION
**Pull request by HenrikSommer**
## Description

With this pull request, we implement `ReadAsync(byte[] data)` with the requested delegate functionality in NuGet package.

The `ReadAsync(byte[] data, MessageType messageType)` in conjunction with the `OnSuccess` and `OnFailure` delegates provides the functionality to deserialize Protobuf data and the flexibility of the receiving domain to handle the received data as desired.

## References

* #588

### CODE REVIEW

File: source/GreenEnergyHub.Charges.Libraries/source/GreenEnergyHub.DataHub.Charges.Clients.Tests/DefaultChargeLink/DefaultChargeLinkReplyDeserializerTests.cs [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    public class DefaultChargeLinkReplyDeserializerTests
+    {
+        private ErrorCode _errorCodeTestResult;
+        private string _knownMeteringPointIdTestResult;
+        private string _unknownMeteringPointIdTestResult;
+        private bool _didCreateChargeLinksTestResult;
+
+        [Theory]
+        [InlineAutoMoqData(MessageType.RequestSucceeded, "knownMeteringPointId1234")]
+        public async Task DeserializeMessageAsync_Test_OnSuccess(MessageType messageType, string meteringPointId)
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges.Libraries/source/GreenEnergyHub.DataHub.Charges.Clients.Tests/DefaultChargeLink/DefaultChargeLinkReplyDeserializerTests.cs:39]: Doesn't seem to comply with out general naming convention (apply to more tests).

File: source/GreenEnergyHub.Charges.Libraries/source/GreenEnergyHub.DataHub.Charges.Clients.Tests/DefaultChargeLink/DefaultChargeLinkReplyDeserializerTests.cs [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+            var target = new DefaultChargeLinkReplyReader(HandleSuccess, HandleFailure);
+
+            // Act
+            await target.ReadAsync(data, messageType).ConfigureAwait(false);
+
+            // Assert
+            target.Should().NotBeNull();
+            _knownMeteringPointIdTestResult.Should().Be(meteringPointId);
+            _didCreateChargeLinksTestResult.Should().BeTrue();
```
[the remark this answers is not in the document]
Comment (HenrikSommer) [reply source/GreenEnergyHub.Charges.Libraries/source/GreenEnergyHub.DataHub.Charges.Clients.Tests/DefaultChargeLink/DefaultChargeLinkReplyDeserializerTests.cs:58]: I will add one

File: source/GreenEnergyHub.Charges.Libraries/source/GreenEnergyHub.DataHub.Charges.Clients.Tests/DefaultChargeLink/DefaultChargeLinkReplyDeserializerTests.cs (hunk 1 above)
Comment (HenrikSommer) [reply source/GreenEnergyHub.Charges.Libraries/source/GreenEnergyHub.DataHub.Charges.Clients.Tests/DefaultChargeLink/DefaultChargeLinkReplyDeserializerTests.cs]: I have used your suggestion (4f9da4f9).

File: source/GreenEnergyHub.Charges.Libraries/source/GreenEnergyHub.DataHub.Charges.Clients.Tests/DefaultChargeLink/Mappers/CreateDefaultChargeLinksFailedInboundMapperTests.cs [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        {
+            // Arrange
+            var createDefaultChargeLinksFailed = _fixture.Create<CreateDefaultChargeLinksFailed>();
+
+            // Act
+            var (meteringPointId, errorCode) = CreateDefaultChargeLinksFailedInboundMapper.Convert(createDefaultChargeLinksFailed);
+
+            // Assert
+            meteringPointId.Should().Be(createDefaultChargeLinksFailed.MeteringPointId);
+            ((int)errorCode).Should().Be((int)createDefaultChargeLinksFailed.ErrorCode);
```
[the remark this answers is not in the document]
Comment (HenrikSommer) [reply source/GreenEnergyHub.Charges.Libraries/source/GreenEnergyHub.DataHub.Charges.Clients.Tests/DefaultChargeLink/Mappers/CreateDefaultChargeLinksFailedInboundMapperTests.cs:46]: Would be a good idea, but would require mapping code in the test, to test mapping in the mapper...

File: source/GreenEnergyHub.Charges.Libraries/source/GreenEnergyHub.DataHub.Charges.Clients.Tests/DefaultChargeLink/Mappers/CreateDefaultChargeLinksFailedOutboundMapperTests.cs [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        {
+            // Arrange
+            var createDefaultChargeLinksFailedDto = _fixture.Create<CreateDefaultChargeLinksFailedDto>();
+
+            // Act
+            var actual = CreateDefaultChargeLinksFailedOutboundMapper.Convert(createDefaultChargeLinksFailedDto);
+
+            // Assert
+            actual.MeteringPointId.Should().Be(createDefaultChargeLinksFailedDto.meteringPointId);
+            ((int)actual.ErrorCode).Should().Be((int)createDefaultChargeLinksFailedDto.errorCode);
```
[the remark this answers is not in the document]
Comment (HenrikSommer) [reply source/GreenEnergyHub.Charges.Libraries/source/GreenEnergyHub.DataHub.Charges.Clients.Tests/DefaultChargeLink/Mappers/CreateDefaultChargeLinksFailedOutboundMapperTests.cs:46]: See comment above

File: source/GreenEnergyHub.Charges.Libraries/source/GreenEnergyHub.DataHub.Charges.Clients/Enums/ErrorCode.cs [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+namespace GreenEnergyHub.DataHub.Charges.Libraries.Enums
+{
+    public enum ErrorCode
+    {
+        ErrorNotDetermined = 0,
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges.Libraries/source/GreenEnergyHub.DataHub.Charges.Clients/Enums/ErrorCode.cs:19]: Perhaps?
```suggestion
        Unspecified = 0,
```

File: source/GreenEnergyHub.Charges.Libraries/source/GreenEnergyHub.DataHub.Charges.Clients/Protobuf/CreateDefaultChargeLinksSucceededOutboundMapper.cs [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+namespace GreenEnergyHub.DataHub.Charges.Libraries.Protobuf
+{
+    internal class CreateDefaultChargeLinksSucceededOutboundMapper
+    {
+        protected internal static CreateDefaultChargeLinksSucceeded Convert(
+            [NotNull] CreateDefaultChargeLinksSucceededDto createDefaultChargeLinksSucceededDto)
+        {
+            if (createDefaultChargeLinksSucceededDto == null)
+                throw new InvalidOperationException();
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges.Libraries/source/GreenEnergyHub.DataHub.Charges.Clients/Protobuf/CreateDefaultChargeLinksSucceededOutboundMapper.cs:28]:
```suggestion
                throw new *ArgumentNullException(nameof(createDefaultChargeLinksSucceededDto));
```

File: source/GreenEnergyHub.Charges.Libraries/source/GreenEnergyHub.DataHub.Charges.Clients/Enums/MessageType.cs [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+namespace GreenEnergyHub.DataHub.Charges.Libraries.Enums
+{
+    public enum MessageType
+    {
+        RequestSucceeded = 1,
+        RequestFailed = 2,
+    }
```
[the remark this answers is not in the document]
Comment (lasrinnil) [reply source/GreenEnergyHub.Charges.Libraries/source/GreenEnergyHub.DataHub.Charges.Clients/Enums/MessageType.cs:21]: +1

**Review by lasrinnil — APPROVED** [review]

File: source/GreenEnergyHub.Charges.Libraries/source/GreenEnergyHub.DataHub.Charges.Clients/Enums/MessageType.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (HenrikSommer) [reply source/GreenEnergyHub.Charges.Libraries/source/GreenEnergyHub.DataHub.Charges.Clients/Enums/MessageType.cs]: Resolved in fb834f1f

**Pull request merged**